README.md
# Workspace Stats

A VSCode extension to provide extra information about workspace, such as biggest files.

## Features

* Workspace Stats: File Sizes

  This command will create a new document where biggest files will be populated across all root folders.

## Extension Settings

This extension contributes the following settings:
* `workspacestats.filesizes.maxPathLen`: Maximum length of the path to populate. Default value = 90.
* `workspacestats.filesizes.minFileSize`: Minimum file size to be shown in the statistics, in bytes. Default value = 10000.

## Known Issues

No issues discovered yet.

## Release Notes

### 0.0.1

Initial release of Workspace Stats.


CHANGELOG.md
# Change Log

## 0.0.1 (2022-Feb-17)

- Initial release.


src/test/suite/fileSvc.test.ts
import * as assert from 'assert';
import * as fileSvc from '../../fileSvc';

suite('File Service Test Suite', () => {
	test('Path.ShrinkPath When_long_path_provided', () => {
		// Arrange
		const path = "C:\\Some Long Path\\With Sub Folders\\And Ends with\\filename.txt";
		const maxLength = 40;

		// Act
		const result = fileSvc.Path.shrinkPath(path, maxLength);

		// Assert
		assert.strictEqual(result, "C:\\Some Long P... Ends with\\filename.txt");
	});

	test('Path.ShrinkPath When_short_path_provided', () => {
		// Arrange
		const path = "C:\\Short path\\filename.txt";
		const maxLength = 40;

		// Act
		const result = fileSvc.Path.shrinkPath(path, maxLength);

		// Assert
		assert.strictEqual(result, "C:\\Short path\\filename.txt");
	});
});


src/fileSvc.ts
import * as fs from 'fs';
import * as vscode from 'vscode';

export interface WorkspaceFolderInfo {
  workspace: string;
  workspacePath: string;
  files: FileInfo[];
}

export interface FileInfo {
  relativeFilePath: string;
  fileName: string;
  ext: string;
  size: number;
  ctime: Date;
  mtime: Date;
}

export class FileSystemProvider {
  static async getAllFiles(minFileSize: number): Promise<WorkspaceFolderInfo[]> {
    if (!vscode.workspace.workspaceFolders)
    {
      return Promise.resolve([]);
    }

    const result: WorkspaceFolderInfo[] = [];
    for (const wf of vscode.workspace.workspaceFolders)
    {
      var foundFiles = await vscode.workspace.findFiles(new vscode.RelativePattern(wf, "**/*.*"));

      const wfi :WorkspaceFolderInfo = {
        workspace: wf.name,
        workspacePath: wf.uri.fsPath,
        files: []
      };

      for (const file of foundFiles)
      {
        const stats = fs.statSync(file.fsPath);

        if (stats.size < minFileSize)
        {
          continue;
        }

        const fileInfo: FileInfo = {
          relativeFilePath: this.getRelativeFilePath(wf, file),
          fileName: this.getFileName(file),
          ext: this.getFileExt(file),
          size: stats.size,
          ctime: stats.ctime,
          mtime: stats.mtime
        };

        wfi.files.push(fileInfo);
      }

      if (wfi.files.length > 0) {
        result.push(wfi);
      }
    }

    return Promise.resolve(result);
  }

  private static getRelativeFilePath(workspaceFolder: vscode.WorkspaceFolder, file: vscode.Uri) {
    return file.fsPath.replace(workspaceFolder.uri.fsPath, "");
  }

  private static getFileName(file: vscode.Uri) {
    return file.fsPath.split(/[/\\]+/).slice(-1)[0];
  }

  private static getFileExt(file: vscode.Uri) {
    return file.fsPath.split('.').slice(-1)[0];
  }
}

export class Path {
  public static shrinkPath(str: string, maxLength: number)
  {
    var splitter = str.indexOf('/')>-1 ? '/' : "\\",
        tokens = str.split(splitter),
        maxLength = maxLength || 25,
        drive = str.indexOf(':')>-1 ? tokens[0] : "",
        fileName = tokens[tokens.length - 1],
        len = drive.length + fileName.length,
        remLen = maxLength - len - 5, // remove the current lenth and also space for 3 dots and 2 slashes
        path, lenA, lenB, pathA, pathB;

        // Remove first and last elements from the array
    tokens.splice(0, 1);
    tokens.splice(tokens.length - 1, 1);

    // Recreate our path
    path = tokens.join(splitter);

    // Handle the case of an odd length
    lenA = Math.ceil(remLen / 2);
    lenB = Math.floor(remLen / 2);

    // Rebuild the path from beginning and end
    pathA = path.substring(0, lenA);
    pathB = path.length > lenB ? path.substring(path.length - lenB) : "";
    let result = "";
    if (drive.length > 0) {
      result = drive + splitter;
    }
    result += pathA;
    if (pathA.length > 0 && pathB.length > 0) {
      result += "...";
    }
    result += pathB;
    if (pathA.length > 0 || pathB.length > 0) {
      result += splitter;
    }
    result += fileName;
    return result;
  }
}


src/extension.ts
import * as vscode from 'vscode';
import * as fileSvc from './fileSvc';

export const EXTENSION_ID = "workspacestats";

export function activate(context: vscode.ExtensionContext) {

	let disposable = vscode.commands.registerCommand(EXTENSION_ID + '.filesizes', async () => {
		const config = vscode.workspace.getConfiguration(EXTENSION_ID);
		const maxPathLenSetting = config.get<number>("filesizes.maxPathLen") ?? 90;
		const minFileSizeSetting = config.get<number>("filesizes.minFileSize") ?? 10000;

		const document = await vscode.workspace.openTextDocument();
		const editor = await vscode.window.showTextDocument(document, 1, false);
		const allFiles = await fileSvc.FileSystemProvider.getAllFiles(minFileSizeSetting);

		let content: string = "";
		for (const wf of allFiles)
		{
			content += `## Biggest files found in '${wf.workspace}':\r\n`;
			let maxPathLen = wf.files.map(x => x.relativeFilePath.length).reduce((p, c) => p > c ? p : c);
			maxPathLen = Math.min(maxPathLenSetting, maxPathLen);
			const maxSizeLen = wf.files.map(x => x.size).reduce((p, c) => p > c ? p : c).toLocaleString().length;
			const filesOrdered = wf.files.sort((a, b) => a.size > b.size ? -1 : (a.size < b.size ? 1 : 0));
			for (const file of filesOrdered) {
				const filepath = fileSvc.Path.shrinkPath(file.relativeFilePath, maxPathLenSetting).padEnd(maxPathLen);
				const filesize = file.size.toLocaleString().padStart(maxSizeLen);
				content += `${filepath}\t${filesize} bytes\r\n`;
			}
		}

		editor?.edit(editBuilder => {
			editBuilder.insert(new vscode.Position(0, 0), content);
		});
	});

	context.subscriptions.push(disposable);
}

export function deactivate() {}
